kubernetes/cache/cache.go

```diff
@@ -0,0 +1,146 @@
+	}
+	// Kiali Cache will use ServiceAccount token instead of user token
+	// Cache creates watchers that have a long cycle to sync with k8s backend maintaining a cache from events
+	// Cache will be used only for *Get* operations, update/delete operations will executed directly against the API
+	// Cache will see what ServiceAccount can see, so when using OpenShift scenarios, user token is used to fetch the
+	// list of projects/namespaces a specific user can see. When using cache, business layer needs to check if a
+	// specific user can see a specific namespace
+	cacheToken := ""
+	kConfig := kialiConfig.Get()
+	if kConfig.InCluster {
+		if saToken, err := kubernetes.GetKialiToken(); err != nil {
+			return nil, err
+		} else {
+			cacheToken = saToken
+		}
+	}
+	istioConfig := rest.Config{
+		Host:            config.Host,
+		TLSClientConfig: config.TLSClientConfig,
+		QPS:             config.QPS,
+		BearerToken:     cacheToken,
+		Burst:           config.Burst,
+	}
+	istioClient, err := kubernetes.NewClientFromConfig(&istioConfig)
+
+	refreshDuration := time.Duration(kConfig.KubernetesConfig.CacheDuration)
+	cacheNamespaces := kConfig.KubernetesConfig.CacheNamespaces
+	kialiCacheImpl := kialiCacheImpl{
+		istioClient:     *istioClient,
+		refreshDuration: refreshDuration,
+		cacheNamespaces: cacheNamespaces,
+		stopChan:        make(chan struct{}),
+		nsCache:     	 make(map[string]typeCache),
+	}
+
+	log.Infof("Kiali Cache is active for namespaces %v", cacheNamespaces)
+	return &kialiCacheImpl, nil
+}
+
+// It will indicate if a namespace should have a cache
+func (c *kialiCacheImpl) isCached(namespace string) bool {
+	for _, cacheNs := range c.cacheNamespaces {
+		if matches, _ := regexp.MatchString(strings.TrimSpace(cacheNs), namespace); matches {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *kialiCacheImpl) createCache(namespace string) bool {
+	informer := make(typeCache)
+	c.createKubernetesInformers(namespace, &informer)
+	c.createIstioInformers(namespace, &informer)
+	c.nsCache[namespace] = informer
+
+	go func() {
+		for _, informer := range c.nsCache[namespace] {
+			go informer.Run(c.stopChan)
+		}
+		<- c.stopChan
+		log.Infof("Kiali cache for [namespace: %s] stopped", namespace)
+	}()
+
+	log.Infof("Waiting for Kiali cache for [namespace: %s] to sync", namespace)
+	isSynced := func () bool {
+		hasSynced := true
+		for _, informer := range c.nsCache[namespace] {
+			hasSynced = hasSynced && informer.HasSynced()
+		}
+		return hasSynced
+	}
+	if synced := cache.WaitForCacheSync(c.stopChan, isSynced); !synced {
+		c.stopChan <- struct{}{}
+		log.Errorf("Kiali cache for [namespace: %s] sync failure", namespace)
+		return false
+	}
+	log.Infof("Kiali cache for [namespace: %s] started", namespace)
+
+	return true
+}
+
+func (c *kialiCacheImpl) CheckNamespace(namespace string) bool {
+	if !c.isCached(namespace) {
+		return false
+	}
+	if _, exist := c.nsCache[namespace]; !exist {
+		return c.createCache(namespace)
```

Comment:
From my testing, you need to have a method to slow down these requests. If too many createCache are called at the same time, you will run into the timeout as the K8s API will throttle the amount of informers that are created. Thus, too many at once will give you timeouts (once there are enough namespaces).

Reply:
Yeah, there are three areas we can touch this:
- Explicit namespaces (only namespaces selected by user are cached).
- Explicit resources (mainly for Istio, but also this can be a flag and let user to decide if he considers some workloads - jobs, replicationcontrollers - I see there are room space here too).
- Also tweak the connection settings, there are some parameters we can also introduce.

And for createCache, one thing can be queue them in a sequential way. In theory that will be a "one time" operation and that would reduce the overhead as well.